users/views/admin/federation.py

```diff
@@ -78,3 +81,58 @@ def get_initial(self):
             "blocked": self.domain.blocked,
             "notes": self.domain.notes,
         }
+
+
+@method_decorator(admin_required, name="dispatch")
+class FederationBlocklist(FormView):
+    template_name = "admin/federation_blocklist.html"
+    extra_context = {"section": "federation"}
+
+    class form_class(forms.Form):
+        blocklist = forms.FileField(
+            help_text=(
+                "Blocklist file with one domain per line. "
+                "Oliphant blocklist format is also supported."
+            ),
+            validators=[FileExtensionValidator(allowed_extensions=["txt", "csv"])],
+        )
+
+    def form_valid(self, form):
+        validator = DomainValidator()
+        domains = []
+
+        try:
+            lines = form.cleaned_data["blocklist"].read().decode("utf-8").splitlines()
+
+            if "#domain" in lines[0]:
+                reader = csv.DictReader(lines)
+            else:
+                reader = csv.DictReader(lines, fieldnames=["#domain"])
+
+            for row in reader:
+                domain = row["#domain"].strip()
+
+                try:
+                    validator(domain)
+                except ValidationError:
+                    # skip adding invalid domain
+                    # to the blocklist
+                    continue
+
+                domains.append(domain)
+        except (TypeError, ValueError):
+            return redirect(".?bad_format=true")
```

Comment:
You can now use the Django messages framework for this instead - `from django.contrib import messages; messages.error(request, "Your CSV was...")`. I switched some of the other error messages so far but not all.

Reply:
Done :slightly_smiling_face: